service/dap/server.go

```diff
@@ -0,0 +1,464 @@
+	// TODO(polina): Respond with an error if debug session is in progress?
+	response := &dap.InitializeResponse{Response: *newResponse(request.Request)}
+	response.Body.SupportsConfigurationDoneRequest = true
+	// TODO(polina): support this to match vscode-go functionality
+	response.Body.SupportsSetVariable = false
+	s.send(response)
+}
+
+func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
+	// TODO(polina): Respond with an error if debug session is in progress?
+	program, ok := request.Arguments["program"]
+	if !ok || program == "" {
+		s.sendErrorResponse(request.Request,
+			3000, "Failed to launch",
+			"The program attribute is missing in debug configuration.")
+		return
+	}
+	s.config.ProcessArgs = []string{program.(string)}
+	s.config.WorkingDir = filepath.Dir(program.(string))
+	// TODO: support program args
+
+	stop, ok := request.Arguments["stopOnEntry"]
+	s.stopOnEntry = (ok && stop == true)
+
+	mode, ok := request.Arguments["mode"]
+	if !ok || mode == "" {
+		mode = "debug"
+	}
+	// TODO(polina): support "debug", "test" and "remote" modes
+	if mode != "exec" {
+		s.sendErrorResponse(request.Request,
+			3000, "Failed to launch",
+			fmt.Sprintf("Unsupported 'mode' value '%s' in debug configuration.", mode))
+		return
+	}
+
+	config := &debugger.Config{
+		WorkingDir:           s.config.WorkingDir,
+		AttachPid:            0,
+		CoreFile:             "",
+		Backend:              s.config.Backend,
+		Foreground:           s.config.Foreground,
+		DebugInfoDirectories: s.config.DebugInfoDirectories,
+		CheckGoVersion:       s.config.CheckGoVersion,
+	}
+	var err error
+	if s.debugger, err = debugger.New(config, s.config.ProcessArgs); err != nil {
+		s.sendErrorResponse(request.Request,
+			3000, "Failed to launch", err.Error())
+		return
+	}
+
+	// Notify the client that the debugger is ready to start accepting
+	// configuration requests for setting breakpoints, etc. The client
+	// will end the configuration sequence with 'configurationDone'.
+	s.send(&dap.InitializedEvent{Event: *newEvent("initialized")})
+	s.send(&dap.LaunchResponse{Response: *newResponse(request.Request)})
+}
+
+// onDisconnectRequest handles the DisconnectRequest. Per the DAP spec,
+// it disconnects the debuggee and signals that the debug adaptor
+// (in our case this TCP server) can be terminated.
+func (s *Server) onDisconnectRequest(request *dap.DisconnectRequest) {
+	s.send(&dap.DisconnectResponse{Response: *newResponse(request.Request)})
+	// TODO(polina): only halt if the program is running
+	if s.debugger != nil {
+		_, err := s.debugger.Command(&api.DebuggerCommand{Name: api.Halt})
+		if err != nil {
+			s.log.Error(err)
+		}
+		kill := s.config.AttachPid == 0
+		err = s.debugger.Detach(kill)
+		if err != nil {
+			s.log.Error(err)
+		}
+	}
+	// TODO(polina): make thread-safe when handlers become asynchronous.
+	s.signalDisconnect()
+}
+
+func (s *Server) onSetBreakpointsRequest(request *dap.SetBreakpointsRequest) {
```

Comment:
One thing to check is if VSCode sends this request while the program is running, if it does know that in that case Debugger.CreateBreakpoint will block until the program stops.

Reply:
> Would there be any issues with the debugger object if I were to call Debugger.CreateBreakpoint in one goroutine and while it was blocking another call to the debugger was made from another goroutine?

Yes, it would block until the other call to the debugger was finished. Debugger serializes all its calls with a mutex, with the exception of the Halt command and `State` with `nowait == true`.